### DISCUSSION TITLE
New module uploaded (keypad)

### DISCUSSION THREAD

**Discussion by salarizadi** [Moddable-OpenSource/moddable#1310] [Announcements]
happy New Year ! I'm back with a new module to set up all kinds of matrix keypads, from now on you can easily set up your own matrix keypads

https://github.com/salarizadi/moddable-sdk/tree/main/Modules/keypad

**Comment by phoddie** [comment]
Very nice. That will be useful for other developers. It also looks like you implement debounce, which is important. You might want to mention that in the docs.

In reading the implementation, I had a couple of minor observations. 

> #pins     = {}

`#pins` is declared as an object but always accessed as an Array. It works, but it would be more clear if the initial value of `#ipns` is an array, `#pins = []`.

```js
    constructor (config, callback) {
        if (typeof config !== "object")
            return false
```
It is a little unusual to have a return value from a constructor. It would be more conventional to throw an exception with an error message.


```js
        if (typeof $this.#pins[config.pin] !== "undefined")
            $this.#pins[config.pin].close()
```

This can be more concisely written as: `$this.#pins[config.pin]?.close()`

**Reply by salarizadi** [reply]
Thanks for reviewing and testing the module

> #pins = {}

If the initial value is an array, accessing the object keys will be impossible and we will get an error. In javascript, I got this error several times when I initialized the array and this small error disrupts the big project, that's why I used the object so that I can easily get the value of the Digital class of the same pin by using the pin number get access. [Be sure to check out this line of code to understand why I used the object](https://github.com/salarizadi/moddable-sdk/blob/69675922cdca93f65eb9ca42a534f1b1760b7033/Modules/keypad/keypad.js#L49])

> if (typeof $this.#pins[config.pin] !== "undefined") 

It makes sense and will be edited as soon as possible

**Reply by phoddie** [reply]
I looked at the line you pointed out, but don't see why that requires an object. Can you give an example of what fails?

By the way, the use of `$this` is not necessary in the `#pin` function. You could achieve the same result by making `onReadable` an arrow function, something like this;

```js
    #pin (config) {
        let which = this.#config.rows.includes(config.pin) ? "rows" : null
        which = this.#config.columns.includes(config.pin) ? "cols" : which

        this.#pins[config.pin]?.close()

        return this.#pins[config.pin] = new Digital({
            edge: Digital.Rising | Digital.Falling,
            ...config,
            onReadable: () => {
                if (which === "cols") return false
                this.#pressed(config, this.#pins[config.pin], which, "click")
            }
        })
    }
```

**Reply by phoddie** [reply]
One more observation. You use `.map()` on the row and column arrays, but always discard the result. In that case, it would be more efficient to use `.forEach()` to avoid creating a new array that is never used. For example, this...

```js
        this.#config.rows.map(pin => this.#pin({
            pin : pin,
            mode: Digital.InputPullUp,
            edge: Digital.Falling
        }))
```

....would become....

```js
        this.#config.rows.forEach(pin => this.#pin({
            pin,
            mode: Digital.InputPullUp,
            edge: Digital.Falling
        }))
```

**Reply by salarizadi** [reply]
I saved the pin settings (digital class) with the object and the #pins keys are the numbers of the pins that have been entered. See this line of code. [Here I got the pin value so that I can perform the operation](https://github.com/salarizadi/moddable-sdk/blob/69675922cdca93f65eb9ca42a534f1b1760b7033/Modules/keypad/keypad.js#L90)

**Reply by salarizadi** [reply]
You mentioned a good point. I rewrote and deleted the codes several times so that I could write this module very accurately and logically. It is normal that some code remains from the previous code that I changed. The things you reported are true, but I want to tell you. Let me explain a part of the previous code: [first look at this line of code](https://github.com/salarizadi/moddable-sdk/blob/69675922cdca93f65eb9ca42a534f1b1760b7033/Modules/keypad/keypad.js#L52)

If you look carefully, the #compressed function is called in this function. Before that, I had written this function in a different way in the onReadable function of the Digital class. That's why I defined the function like this. Instead of () => {}, I wanted to directly have this.read() to get the pin value, but I saw that I needed more operations, so I went back and changed the codes. Now you understand why some code doesn't make sense. Yes, I could have used this instead of $this. I wrote the codes without sample code, I changed several times to be able to identify the keys exactly. It sounds simple, but it's a bit complicated.

**Reply by salarizadi** [reply]
It is interesting that I had not paid attention to this until now. Using map creates a new array. This can also be changed and I change it to forEach. Thanks for the tips you reported

**Reply by salarizadi** [reply]
Now that I paid attention, I was able to write the codes much better and lighter. I will definitely update the codes as soon as possible. I really appreciate your detailed review.
